### ISSUE TITLE
CPU execution of ops after gradient clipping on windows

### ISSUE THREAD

**Issue by Dhruv-Mohan** [tensorflow/tensorflow#16864]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows10
- **TensorFlow installed from (source or binary)**: Binary
- **TensorFlow version (use command below)**:1.5.0
- **Python version**: 3.6
- **Bazel version (if compiling from source)**:N/A
- **GCC/Compiler version (if compiling from source)**:N/A
- **CUDA/cuDNN version**:CUDA 9.0, cuDNN 7
- **GPU model and memory**: Titan xp, 12gb
- **Exact command to reproduce**: Script attached below

### Describe the problem
Adding gradient clipping as follows places certain ops on the CPU greatly increasing training time on windows.
```python
    grads = tf.gradients(loss, tf.trainable_variables())
    omnomed_grads, _ = tf.clip_by_global_norm(grads, 0.5)
    train_op = optimizer.apply_gradients(zip(omnomed_grads,  tf.trainable_variables()))
```
I worte a [script](https://github.com/Dhruv-Mohan/G_clip_issue/blob/master/Gptest.py)  to perform a few tests:

| OS        | TF Version           | Gradient clipping  | Average runtime (m:s.ms) |
| :-------------: |:-------------:| :-----:| :--------: |
| Windows10           | 1.6  | True   | 2:13.57    |
| Windows10           | 1.6  |  False | 0:16.24    |
| Windows10           | 1.5  | True   | 2:13.46    |
| Windows10           | 1.5  |  False | 0:16.52    |
| Windows10           | 1.2  |  True  | 0:24.64    |
| Windows10           | 1.2  |  False | 0:23.80    |
| Linux mint 18.1     | 1.5  | True   | 0:23.45    |
|Linux mint 18.1      | 1.5  |  False | 0:21.29    |

Curiously the issue isn't present on TF1.2

Forcing operation placement on the GPU:
```python
    with tf.device('/gpu:0'):
        grads = tf.gradients(loss, tf.trainable_variables())
        omnomed_grads, _ = tf.clip_by_global_norm(grads, 0.5)
        train_op = optimizer.apply_gradients(zip(omnomed_grads,  tf.trainable_variables()))
```
results in the following error on windows:
 `Cannot assign a device for operation 'gradients/softmax_cross_entropy_with_logits_grad/LogSoftmax': Could not satisfy explicit device specification '/device:GPU:0' because no supported kernel for GPU devices is available.`
The error generated is similar to [#2803](https://github.com/tensorflow/models/issues/2803) and [#3118](https://github.com/tensorflow/models/issues/3118)

Setting soft device placement in session configuration results in similar runtime.

### Source code / logs
I have attached the device placement logs of windows 10 running TF 1.6 [with](https://github.com/tensorflow/tensorflow/files/1707039/Windows_gradclip.txt) and [without](https://github.com/tensorflow/tensorflow/files/1707040/Windows_nogradclip.txt) gradient clipping as well as the device placement logs from Linux mint [with](https://github.com/tensorflow/tensorflow/files/1707038/Mint_gradclip.txt) gradient clipping

A few discrepancies with respect to op placement while gradient clipping are highlighted below:

| Windows10        |  Linux mint        |
| :-------------: |:-------------:| 
| gradients/softmax_cross_entropy_with_logits_grad/LogSoftmax: (LogSoftmax): /job:localhost/replica:0/task:0/device:CPU:0     | gradients/softmax_cross_entropy_with_logits_grad/LogSoftmax: (LogSoftmax): /job:localhost/replica:0/task:0/device:GPU:0 | 
| global_norm/L2Loss_6: (L2Loss): /job:localhost/replica:0/task:0/device:CPU:0      | global_norm/L2Loss_6: (L2Loss): /job:localhost/replica:0/task:0/device:GPU:0      | 
|gradients/conv2d/Conv2D_grad/Conv2DBackpropFilter: (Conv2DBackpropFilter): /job:localhost/replica:0/task:0/device:CPU:0 | gradients/conv2d/Conv2D_grad/Conv2DBackpropFilter: (Conv2DBackpropFilter): /job:localhost/replica:0/task:0/device:GPU:0      | 
| gradients/conv2d/BiasAdd_grad/BiasAddGrad: (BiasAddGrad): /job:localhost/replica:0/task:0/device:CPU:0   | gradients/conv2d/BiasAdd_grad/BiasAddGrad: (BiasAddGrad): /job:localhost/replica:0/task:0/device:GPU:0   |
| clip_by_global_norm/mul_8: (Mul): /job:localhost/replica:0/task:0/device:CPU:0 | clip_by_global_norm/mul_8: (Mul): /job:localhost/replica:0/task:0/device:GPU:0 |

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by adler-j** [comment]
Did you ever make any progress with this? I've had the same issue myself and it seems to be exactly this happening.

**Comment by Dhruv-Mohan** [comment]
Sorry, haven't made any progress on this issue. I switched training of networks to Linux to mitigate the problem.
________________________________

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by George-Gate** [comment]
I also have this problem, it seems that there are only CPU kernels for solfmax...
Code:
![image](https://user-images.githubusercontent.com/8580553/38312187-a7c93f52-3853-11e8-92a7-d6a0a2dac2c2.png)
Result:
![image](https://user-images.githubusercontent.com/8580553/38312131-886bc3d2-3853-11e8-917c-cac4c3fd4a58.png)

I am using tensorflow-1.7.0-rc1 (GPU version from pip) on Windows 10, with GTX 1080Ti card. This is bad  for me since all other ops of my graph is on GPU......

**Comment by George-Gate** [comment]
A tradeoff is to define your own softmax:
```
def my_softmax(logits, axis = -1):
    tmp = tf.exp(logits)
    return tmp / tf.reduce_sum(tmp, axis, keepdims = True)
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 52 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 66 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by adler-j** [comment]
Is there any chance for an update on this? This bug is basically rendering a lot of my old code un-usable since it takes weeks to run it. And dropping the clipping changes its behavior.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 62 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 78 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 94 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 108 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 123 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 138 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 154 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 169 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mohantym** [comment]
Hi @Dhruv-Mohan ! 1.x versions are not supported any more. Have you checked this issue in 2.7 version onward? Thanks!
